### PR TITLE
Button Widget: Uniform Spacing

### PR DESCRIPTION
**Pull request by tanishqmanuja**
## Summary

Button widget had asymmetrical top and bottom spacing. This PR tries to fix that.
Related issue: https://github.com/home-assistant/android/issues/3093

## Screenshots

Portrait:
image

Landscape:
image

## Link to pull request in Documentation repository

Documentation: home-assistant/companion.home-assistant#

## Any other notes

- The icon named old is from stock Home Assistant App kept for comparison with the newer ones.
- The only small downside of this approach is that on multiline text the icon becomes slightly smaller as compared to stock multi line icon (BTW stock multiline widget is not present in the screenshots above)

In order to fully solve this only downside with even better but complex approach is to use multiple layouts:
https://developer.android.com/develop/ui/views/appwidgets/layouts#provide-responsive-layouts

### CODE REVIEW

**Comment by jpelgrom** [comment]
This still isn't quite equal spacing, unless I'm missing something. The root view `widgetLayout` has 4dp padding all around, and the label view `widgetLabelLayout` has 2dp padding all around, so in total there is now 4dp padding at the top but 6dp padding at the bottom.

There is also an additional 4dp vertical padding for `widgetImageButtonLayout` - why is that? It contributes to the smaller icon you're describing.

The entity state widget uses very similar spacing for it's elements, if the button widget is updated I feel like that one should be too in this PR to keep their layouts aligned.

In combination with your previous PR, should this close the issue?

**Comment by tanishqmanuja** [comment]
Maybe not perfectly symmetrical (with difference of 2dp that too only on some devices).

Top Spacing -> `4dp` from Widget Layout + `4dp` from Vertical padding of Linear Layout
Bottom Spacing -> `4dp` from Widget Layout + `4dp` from Vertical padding of Linear Layout + `2dp` from margin of Label Layout (to protect text character like 'j', 'g' which in my experience leaks on user fonts in Custom ROMs)

TLDR:
 - So the original 4dp from Widget Layout was already there before my PR, so I didn't touch it. I added extra 4dp to make total 8dp which is the `Widget Corner Radius`.
- `In combination with your previous PR, should this close the issue?` Yess :)

PS:
2dp not adding to bottom spacing when enough space is provided according to layout bounds.
image

**Comment by tanishqmanuja** [comment]
@jpelgrom, Regarding the entity state widget what should I do ?
BTW in my testing I found stock widget is already broken in multiline scenarios for landscape mode.

**Comment by tanishqmanuja** [comment]
Commit https://github.com/home-assistant/android/pull/3099/commits/f3a964e031aefa806b74f838949f1553d228027a,
Mirror the changes from button widget to entity state widget.

Screenshot:
image

**Comment by jpelgrom** [comment]
> I added extra 4dp to make total 8dp which is the `Widget Corner Radius`.

The widget corner radius, especially when using the Dynamic color theme like you do, will depend on the widget theme, Android version, launcher, device and/or manufacturer. If you want to use the widget corner radius (which can be >8dp), you can use `?dimenWidgetCornerRadius`.

> Regarding the entity state widget what should I do ?

I think changes to mirror the button widget are good 👍

> I found stock widget is already broken in multiline scenarios for landscape mode

There is a big chance it will be too small when using a modern phone, but for less tall devices / tablets it is less of a problem. Right now the text size is configurable by users so it's not easily solved without adding more options, changing it to a less exact option like 'small/medium/large' or adding a completely different layout, which is IMO not in the scope for this PR.

(Adding more padding like you've done _does_ increase the chances of there not being enough space for the text.)

**Comment by tanishqmanuja** [comment]
So is the PR complete or something more should be done?

File: app/src/main/res/layout/widget_button.xml [hunk 1]
```diff
@@ -31,18 +32,17 @@
...
                 android:id="@+id/widgetLabel"
                 android:layout_width="match_parent"
                 android:layout_height="wrap_content"
-                android:layout_gravity="center"
+                android:layout_gravity="bottom|center_horizontal"
                 android:text="@string/widget_label_placeholder_text_button"
                 android:textSize="12sp"
                 android:textAlignment="center"
-                android:gravity="center"
+                android:gravity="top|center_horizontal"
```
Comment (jpelgrom) [inline app/src/main/res/layout/widget_button.xml:45]: As the TextView height is `wrap_content`, aligning it to the top doesn't make any difference so this line doesn't need to be changed.

File: app/src/main/res/layout/widget_button.xml [hunk 2]
```diff
@@ -31,18 +32,17 @@
...
-            android:layout_margin="2dp"
-            android:minHeight="32sp">
+            android:layout_margin="2dp">
 
             <TextView
                 android:id="@+id/widgetLabel"
                 android:layout_width="match_parent"
                 android:layout_height="wrap_content"
-                android:layout_gravity="center"
+                android:layout_gravity="bottom|center_horizontal"
```
Comment (jpelgrom) [inline app/src/main/res/layout/widget_button.xml:41]: As the parent LinearLayout height is `wrap_content` and no longer has a minimum height, aligning it to the bottom doesn't make any difference so this line doesn't need to be changed.

File: app/src/main/res/layout/widget_button.xml (hunk 1 above)
Comment (tanishqmanuja) [reply app/src/main/res/layout/widget_button.xml]: Thanks for mentioning

**Review by jpelgrom — APPROVED** [review]

**Comment by jpelgrom** [comment]
Not uniform but close enough when there is a single line of text 😉

**Review by JBassett — APPROVED** [review]
